### PR TITLE
Defer tweet enhancement

### LINKED ISSUE #4620: Defer tweet enhancement

### PR DESCRIPTION
**Pull request by oliverlloyd**
## What does this change?
Defers enhancement of Twitter tweets until they scroll into view.  Fixes #4620 

This work matches what was [originally](https://github.com/guardian/frontend/pull/6203) [implemented](https://github.com/guardian/frontend/pull/6244) in Frontend

Previously we were using a `script` tag on twitter embeds which [loaded the twitter javascript](https://developer.twitter.com/en/docs/twitter-for-websites/embedded-tweets/overview) for each tweet. When this code was loaded and run it found each tweet on the page and 'enhanced' it.

We now [centralise the loading](https://developer.twitter.com/en/docs/twitter-for-websites/javascript-api/guides/set-up-twitter-for-websites) of this javascript and have [more control](https://developer.twitter.com/en/docs/twitter-for-websites/javascript-api/guides/scripting-loading-and-initialization) over when the enhancement happens.

## Why?
To improve page load experience, esp. for people using permalinks

## How are we deferring the `load` action?
Twitter will automatically run it's enhancement js on all tweets that have the `.twitter-tweet` class. By removing this class at the point we get the raw html from Frontend (using an 'enhancer') we prevent this happening. We later only add this classname back in when the element is visible and then run `twitter.widgets.load` on just that part of the page

### `enhance` and `enhance`?
Sadly, there are two meanings to the word enhance in this PR. The Twitter meaning, which is to take a basic embedded tweet and to add more features, images, tracking, etc. to it and then DCR's meaning, which is where we have a series of 'enhance' functions that take raw Frontend data and transform it into something more useable for DCR.

### CODE REVIEW

File: dotcom-rendering/src/model/enhance-tweets.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+				const withoutTweetClass = element.html.replace(
+					'twitter-tweet',
+					'nojs-tweet',
+				);
+				enhanced.push({
+					...element,
+					...{
+						html: withoutTweetClass,
+					},
+				});
```
Comment (jamesgorrie) [inline dotcom-rendering/src/model/enhance-tweets.ts:18]: not a biggy, but maybe easier to read
```suggestion ts
				enhanced.push({
					...element,
					html: withoutTweetClass,
				});
```

File: dotcom-rendering/src/web/components/TweetBlockComponent.importable.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		/* stylelint-disable-next-line declaration-no-important */
+		width: 100% !important;
+	}
+
+	.nojs-tweet p {
+		padding-bottom: 10px;
+	}
+
+	a {
+		/* stylelint-disable-next-line color-no-hex */
```
Comment (jamesgorrie) [inline dotcom-rendering/src/web/components/TweetBlockComponent.importable.tsx:32]: ? Is this a twitter colour?

**Review by jamesgorrie — APPROVED** [review]
This is a very elegant solution!

**Comment by jamesgorrie** [comment]
Bikeshed re: enhance and enhance - why do we use `enhance` over `transform`? It just seems like we don't add much, rather just change. 

I think I get twitters reason that you have a plain tweet with noJS and a fancy one with JS, thus enhanced.

File: dotcom-rendering/src/web/components/TweetBlockComponent.importable.tsx (hunk 2 above)
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/TweetBlockComponent.importable.tsx:32]: No idea 🤷  - this is isn't my code, I renamed the file so it got green'd

File: dotcom-rendering/src/web/server/articleToHtml.tsx [hunk 3]
```diff
@@ -193,5 +219,9 @@ export const articleToHtml = ({ data }: Props): string => {
 		openGraphData,
 		twitterData,
 		keywords,
+		initTwitter:
+			pageHasTweetElements || format.design === ArticleDesign.LiveBlog
```
[the remark this answers is not in the document]
Comment (oliverlloyd) [reply dotcom-rendering/src/web/server/articleToHtml.tsx:223]: Yes, basically. I think there's a similar preference in Frontend to - kind of - expect tweet embeds in blogs.

The alternative is to not load the javascript here and defer it to when (and if) a tweet appears. The downside to this is a slower experience for that tweet block and more complex code.

File: dotcom-rendering/src/model/enhance-tweets.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+const removeTweetClass = (elements: CAPIElement[]): CAPIElement[] => {
+	const enhanced: CAPIElement[] = [];
+	elements.forEach((element) => {
```
Comment (mxdvl) [inline dotcom-rendering/src/model/enhance-tweets.ts:3]: If we mutate, I think `for … of` is more explicit

**Comment by oliverlloyd** [comment]
> Bikeshed re: enhance and enhance - why do we use `enhance` over `transform`? It just seems like we don't add much, rather just change.
> 
> I think I get twitters reason that you have a plain tweet with noJS and a fancy one with JS, thus enhanced.

I guess one word makes more sense over another at a point in time and perhaps over time things change? I know in Frontend the same sorts of actions are called 'cleaners' which is an interesting choice. Maybe we wanted to be more positive alongside this?

File: dotcom-rendering/src/web/components/TweetBlockComponent.importable.tsx (hunk 2 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/components/TweetBlockComponent.importable.tsx:32]: This is somewhat odd… usually git managed to see this as a rename :(

File: dotcom-rendering/src/web/lib/renderElement.tsx [hunk 5]
```diff
@@ -600,6 +600,14 @@ export const renderElement = ({
...
 		case 'model.dotcomrendering.pageElements.TweetBlockElement':
+			if (switches.enhanceTweets) {
+				return [
+					true,
+					<Island deferUntil="visible">
+						<TweetBlockComponent element={element} />
+					</Island>,
+				];
+			}
 			return [true, <TweetBlockComponent element={element} />];
```
Comment (mxdvl) [inline dotcom-rendering/src/web/lib/renderElement.tsx:611]: Wouldn’t that get two tweets?

File: dotcom-rendering/src/web/server/articleToHtml.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/web/server/articleToHtml.tsx:223]: That’s how I read this

**Review by mxdvl — APPROVED** [review]
Nice

File: dotcom-rendering/src/web/lib/renderElement.tsx (hunk 5 above)
Comment (oliverlloyd) [reply dotcom-rendering/src/web/lib/renderElement.tsx:611]: It's returning, so you can only get one. My linter insists on this pattern.

**Comment by jamesgorrie** [comment]
> cleaners

I suppose the thinking was it only ever removed stuff.

> Maybe we wanted to be more positive alongside this?

True, but it's nice to try use language others will understand quickly, but, as usual that's hard, especially as things change.

**Pull request merged**